dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/ContextBeginnerEnderImpl.java

```diff
@@ -0,0 +1,117 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.liberty;
+
+import com.ibm.ws.cdi.CDIException;
+import com.ibm.ws.cdi.internal.interfaces.Application;
+import com.ibm.ws.cdi.internal.interfaces.CDIArchive;
+import com.ibm.ws.cdi.internal.interfaces.CDIUtils;
+import com.ibm.ws.cdi.internal.interfaces.ContextBeginnerEnder;
+import com.ibm.ws.runtime.metadata.ApplicationMetaData;
+import com.ibm.ws.runtime.metadata.MetaData;
+import com.ibm.ws.runtime.metadata.ModuleMetaData;
+import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
+
+public class ContextBeginnerEnderImpl implements ContextBeginnerEnder {
+
+    private ClassLoader newTheadContextClassLoader = null;
+    private ClassLoader oldTheadContextClassLoader = null;
+    private JndiHelperComponentMetaData cmd = null;
+    private final boolean began = false;
+
```

Comment:
There are several things I don't like about this class.

1. The scope is wrong: this shouldn't be a singleton that's stored in `CDIRuntime` because it holds state which relates to the latest extracted and applied context. Therefore you should be making a new object every time you extract context.
2. Two different jobs are conflated. This object holds both the new TCCL (i.e. the one to apply) and the old TCCL (i.e. the one to restore when the context is ended). This should be two classes, one which olds the new TCCL, created when context is extracted, and one which holds old TCCL, created when context is applied (and which can be used to remove the context and restore the original context).
    - The second class should implement `AutoClosable` and where possible you should call it with a try-with-resources block, ensuring that the context applied at the start of the block is always removed at the end.